### ISSUE TITLE
Retrofit 2 with RxJava - combine.zip issue

### ISSUE THREAD

**Issue by dastko** [square/retrofit#2031]
Hi,

I'm having an issue with RxJava and Retrofit (tested with ZIP, COMBINE or COMBINE LATEST function). 
I'm using the ZIP method to combine multiple API calls and the result should be several arrays with some data (server always returns the same set of data).
The problem I have is that sometimes some arrays are empty (in onNext/onCompleted methods) even if the server returns data.
It happens approximately every 10th call. 
The code snippet is:

`
private void getFullClientDetails() {

```
    repo = ServiceGenerator.createServiceRx(ApplicationController.getInstance(), mApplicationSettings, this);
    ProgressDialogUtil.showProgressDialog(ClientBaseActivity.this);
    ScreenOrientationLock.lockScreenOrientation(this);
    Observable<ArrayList<Assessment>> assessmentObservable = getAssessmentObservable(repo);
    Observable<ArrayList<Report>> reportObservable = getReportObservable(repo);
    Observable<ArrayList<CarePlan>> carePlanObservable = getCarePlanObservable(repo);
    Observable<ArrayList<Evaluation>> evaluationObservable = getEvaluationObservable(repo);
    Observable<ArrayList<GuidanceAppointment>> appointmentObservable = getAppointmentObservable(repo);

    final Observable<MedicalInfoObservable> combined = Observable.zip(assessmentObservable, reportObservable, carePlanObservable, evaluationObservable, appointmentObservable,
            new Func5<ArrayList<Assessment>,
                    ArrayList<Report>, ArrayList<CarePlan>, ArrayList<Evaluation>, ArrayList<GuidanceAppointment>, MedicalInfoObservable>() {
                @Override
                public MedicalInfoObservable call(ArrayList<Assessment> assessments,
                                                  ArrayList<Report> reports, ArrayList<CarePlan> carePlan, ArrayList<Evaluation> evaluations, ArrayList<GuidanceAppointment> appointments) {
                    return new MedicalInfoObservable(assessments, reports, carePlan, evaluations, appointments);
                }
            });


    subscription = combined.subscribe(new Subscriber<MedicalInfoObservable>() {
        @Override
        public void onCompleted() {
            openDetailsTabActivity();
        }

        @Override
        public void onError(Throwable e) {
            showErrorMessage(e);
        }

        @Override
        public void onNext(MedicalInfoObservable medicalInfoObservable) {

                mAssessmentResponse = (ArrayList<Assessment>) medicalInfoObservable.assessmentList;
                mReportsResponse = (ArrayList<Report>) medicalInfoObservable.reportList;
                mCarePlanList = (ArrayList<CarePlan>) medicalInfoObservable.carePlanList;
                mEvaluationsList = (ArrayList<Evaluation>) medicalInfoObservable.evaluationList;
                mAppointmentList = (ArrayList<GuidanceAppointment>) medicalInfoObservable.appointmentList;
                Log.e("ClientObservable", "Appointment List: " + medicalInfoObservable.appointmentList.size() +
                        ", Assessment List: " + medicalInfoObservable.assessmentList.size() + ",Care Plan: " + medicalInfoObservable.carePlanList.size() +
                        ", Evaluation List: " + medicalInfoObservable.evaluationList.size() + ",Report" + medicalInfoObservable.reportList.size()); 
                        }


    });
}



private Observable<ArrayList<Assessment>> getAssessmentObservable(Retrofit repo) {
    return repo
            .create(AssessmentApi.class)
            .getAssessmentListRx(mCurrentlySelectedClient.getId(), true)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread());
}

private Observable<ArrayList<Report>> getReportObservable(Retrofit repo) {
    return repo
            .create(ReportApi.class)
            .getReportRxWithClient(mCurrentlySelectedClient.getId(), true)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread());
}

private Observable<ArrayList<CarePlan>> getCarePlanObservable(Retrofit repo) {
    return repo
            .create(CarePlanApi.class)
            .getCarePlanListRx(mCurrentlySelectedClient.getId(), true)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread());
}

private Observable<ArrayList<Evaluation>> getEvaluationObservable(Retrofit repo) {
    return repo
            .create(EvaluationApi.class)
            .getEvaluationListRx(mCurrentlySelectedClient.getId(), true)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread());
}

private Observable<ArrayList<GuidanceAppointment>> getAppointmentObservable(Retrofit repo) {
    return repo
            .create(GuidanceAppointmentApi.class)
            .getGuidanceAppointmentListRx(mCurrentlySelectedClient.getId(), true)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread());
}

@Override
protected void onDestroy() {
    super.onDestroy();
    if(subscription != null)
        subscription.unsubscribe();
}
```

`
Response from the server is always the same (**7,4,30** items in list), but in onNext/onCompleted method I have different results. The example is:

Appointment List: 7, Assessment List: 4,Care Plan: 30, Evaluation List: 0,Report 0
Appointment List: 7, Assessment List: 4,Care Plan: 30, Evaluation List: 0,Report 0
Appointment List: 7, Assessment List: 4,Care Plan: 30, Evaluation List: 0,Report 0
Appointment List: 0, Assessment List: 4,Care Plan: 30, Evaluation List: 0,Report 0
Appointment List: 7, Assessment List: 4,Care Plan: 0, Evaluation List: 0,Report 0

One more example:
Appointment List: 7, Assessment List: 4,Care Plan: 30, Evaluation List: 0,Report 0
Appointment List: 7, Assessment List: 4,Care Plan: 30, Evaluation List: 0,Report 0
Appointment List: 7, Assessment List: 4,Care Plan: 30, Evaluation List: 0,Report 0
Appointment List: 30, Assessment List: 4,Care Plan: 30, Evaluation List: 0,Report 0
Appointment List: 7, Assessment List: 4,Care Plan: 30, Evaluation List: 0,Report 0
Appointment List: 7, Assessment List: 4,Care Plan: 30, Evaluation List: 0,Report 0
Appointment List: 7, Assessment List: 4,Care Plan: 30, Evaluation List: 7,Report 4

Build gradle dependencies: 
    compile 'com.squareup.retrofit2:retrofit:2.1.0'
    compile 'com.squareup.retrofit2:converter-gson:2.1.0'
    compile 'io.reactivex:rxandroid:1.2.1'
    compile 'io.reactivex:rxjava:1.1.8'
    compile 'com.squareup.retrofit2:adapter-rxjava:2.1.0'
    compile 'com.squareup.okhttp3:logging-interceptor:3.4.1'

If you need more information just let me now.

**Comment by JakeWharton** [comment]
Can you express this in a failing test? Preferably one without so much machinery going on. Without a way to reproduce the behavior there's no way to know whether this is a Retrofit problem, RxJava problem, or user error.

**Comment by dastko** [comment]
Hi @JakeWharton ,

I also reported an issue on RxJava channel and after deep testing and multiple scenarios, this problem is related to Retrofit, not RxJava.
Probably Retrofit mixes up the data between requests because of reusing the same connection and not properly demuxing.

For example, sometimes I got the list of Evaluation (or Report, there is no a clear pattern) objects with properties that belong to Care Plan object (the common property for this two objects is id) and customers can see the list of objects that doesn't exist in the database.

I've tried with one suggestion:

ExecutorService sc = Executors.newFixedThreadPool(1);
scheduler = Schedulers.from(sc);

Everything works just fine, but with one thread in the pool. When I tried with 4 or more threads in the pool - issue appears.
